### DISCUSSION TITLE
ImGui InputText - Backspace and Arrow Keys not Functional with osgEarth Integration

### DISCUSSION THREAD

**Discussion by nebsar** [pelicanmapping/osgearth#1973] [Support Q&A]
Hello,

Backspace and left and right arrow keys are not functional on Imgui InputText widget with osgEarth integration. What could be the reason?

**Comment by gwaldron** [comment]
I'm not able to reproduce your issue using the Search tool in osgearth_imgui.cpp Backspace and the Arrows appear to be working.

**Reply by nebsar** [reply]
Yes Glenn, The InputText works fine with osgearth_imgu example when the ImGui docker 1.84 is used. But I think there is a problem with ImGui 1.87 which is the last version, as I see. Would you please try it as well?

**Comment by nebsar** [comment]
Hi Glenn,

I am using the osgearth_imgui.cpp. I do not have geocoder so I do not use the search tool. But in ImGui demo window InputText backspace and arrow keys does not seem to work at least when I run the example. Here is a video showing that I can only  edit the text only by selecting it and typing something with keyboard.

Edit: I am using the master branch source code of osgEarth in the github and Imgui version 1.87 docking which is the last version in the github.

https://user-images.githubusercontent.com/5169314/153871507-d4494611-c72d-49ee-9af5-6741fa5f5851.mp4

**Reply by jasonbeverage** [reply]
osgearth should be getting imgui from a pinned submodule git sha so the version that is in there should work.  We haven't tried updating to the latest imgui yet, so it is good to note that we might run into issues.  If you want to take a stab at figuring out what is wrong with the latest imgui and submit a PR or a report that would be awesome.

**Reply by nebsar** [reply]
Hi Jason,
I found where the issue is, for the newer version of the ImGui. It is about key mapping. Inside the imgui.h, the enum ImGuiKey_ is completely different. And the key array is not less than 512.
There is also problem with mouse clicks. Especially for mouse double click.

Here is the explanation from imgui.h:

  // Legacy: before 1.87, we required backend to fill io.KeyMap[] (imgui->native map) during initialization and io.KeysDown[] (native indices) every frame.
    // This is still temporarily supported as a legacy feature. However the new preferred scheme is for backend to call io.AddKeyEvent().
#ifndef IMGUI_DISABLE_OBSOLETE_KEYIO
    int         KeyMap[ImGuiKey_COUNT];             // [LEGACY] Input: map of indices into the KeysDown[512] entries array which represent your "native" keyboard state. The first 512 are now unused and should be kept zero. Legacy backend will write into KeyMap[] using ImGuiKey_ indices which are always >512.
    bool        KeysDown[ImGuiKey_COUNT];           // [LEGACY] Input: Keyboard keys that are pressed (ideally left in the "native" order your engine has access to keyboard keys, so you can use your own defines/enums for keys). This used to be [512] sized. It is now ImGuiKey_COUNT to allow legacy io.KeysDown[GetKeyIndex(...)] to work without an overflow.
#endif

**Reply by jasonbeverage** [reply]
Great!  Thanks for looking into this.  We'll use this information next time we upgrade imgui in osgEarth.

**Comment by nebsar** [comment]
Hi Jason,

I will try to do that. Thanks
